services/src/main/java/org/keycloak/services/DefaultKeycloakSessionFactory.java

```diff
@@ -117,25 +118,56 @@ public void init() {
                 }
             }
             checkProvider();
-            // Component factory must be initialized first, so that postInit in other factories can use component factories
-            updateComponentFactoryProviderFactory();
-            if (componentFactoryPF != null) {
-                componentFactoryPF.postInit(this);
-            }
-            for (Map<String, ProviderFactory> factories : factoriesMap.values()) {
-                for (ProviderFactory factory : factories.values()) {
-                    if (factory != componentFactoryPF) {
-                        factory.postInit(this);
-                    }
-                }
-            }
+            initProviderFactories();
             // make the session factory ready for hot deployment
             ProviderManagerRegistry.SINGLETON.setDeployer(this);
         }
 
         AdminPermissions.registerListener(this);
     }
 
+    protected void initProviderFactories() {
+        // Component factory must be initialized first, so that postInit in other factories can use component factories
+        updateComponentFactoryProviderFactory();
+        if (componentFactoryPF != null) {
+            componentFactoryPF.postInit(this);
+        }
```

Comment:
Never fully understood what component factories where, and why they needed to be initialised early. Seems that really this could/should be refactored to use dependencies as well?

Reply:
I can experiment with removing this and see what happens without calling this first in all cases. Worst case, assuming this breaks a lot of things, a simple solution would be to make the default `ProviderFactory#dependsOn` method return a singleton set `ComponentFactoryProviderFactory`. Subsequent implementations could then add to this set, whereas the `ComponentFactoryProviderFactory#dependsOn` implementation would return an empty set as it obviously can't depend on itself.

Reply:
Actually not sure AmphibianProviderFactory or ComponentFactoryProvider is even used anymore. It was something that was introduced for the new store, but now that's gone I don't think there's anything using it anymore.

Reply:
^ is not possible due to `ComponentFactoryProviderFactory` being part of the `keycloak-server-spi-private` module and `ProviderFactory` being in `keycloak-server-spi`.